vllm/v1/core/scheduler.py

```diff
@@ -330,6 +367,23 @@ def schedule(self) -> "SchedulerOutput":
                 self.kv_cache_manager.get_num_common_prefix_blocks(
                     any_request, len(self.running)))
 
+        # Prepare the guided decoding bitmask for this batch.
+        grammar_bitmask: Optional[np.ndarray] = None
+        if guided_decoding_request_ids:
+            # Fill the bitmask using the index of each request equal to its
+            # position in the batch. Resize the bitmask down to the size of
+            # the batch.
+            bitmask_tensor = self.guided_decoding_manager.grammar_bitmask
+            assert bitmask_tensor is not None
+            for req_id, batch_index in guided_decoding_request_ids.items():
+                request = self.requests[req_id]
+                assert request.grammar is not None
+                if not request.grammar.matcher.is_terminated():
+                    request.grammar.fill_bitmask(bitmask_tensor, batch_index)
```

Comment:
I'm a bit confused. If this is the purpose of `guided_decoding_request_ids: Dict[str, int]`, can we simply use `List[str]`?

Reply:
The `int` part is needed in the GPU model runner, as well. It needs to know which index in the bitmask corresponds to which request.

Reply:
The fundamental issue is that the batch ordering is not the same here as in the GPU model runner. If it was, we could use `List[str]`. Instead we have to re-order the bitmask to match the batch ordering in the model runner, but we need this info to know how it was ordered when sent from the scheduler.